boot/seal.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if err := secbootSealKey(key, &sealKeyParams); err != nil {
+		return fmt.Errorf("cannot seal the encryption key: %v", err)
+	}
+
+	return nil
+}
+
+// recoverModeLoadSequences builds the recovery boot file list based on
+// the boot chain retrieved from the bootloader and trusted assets listed
+// in modeenv.
+func recoverModeLoadSequences(rbl bootloader.Bootloader, modeenv *Modeenv) ([][]bootloader.BootFile, error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("bootloader doesn't support trusted assets")
+	}
+
+	kernelPath, err := kernelPathForRecoverySystem(modeenv.RecoverySystem)
+	if err != nil {
+		return nil, err
+	}
+
+	bootChain, err := tbl.RecoveryBootChain(kernelPath)
+	if err != nil {
+		return nil, err
+	}
+
+	seq0 := make([]bootloader.BootFile, 0, len(bootChain))
+	seq1 := make([]bootloader.BootFile, 0, len(bootChain))
+
+	// walk the chain and get cache entries for the trusted assets
+	for i := 0; i < len(bootChain)-1; i++ {
+		bf := bootChain[i]
+		name := filepath.Base(bf.Path)
+
+		if bf.Role != bootloader.RoleRecovery {
+			return nil, fmt.Errorf("internal error: recovery boot chain contains invalid role %v", bf.Role)
+		}
+
+		p0, p1, err := cachedAssetPathnames(rbl.Name(), name, modeenv.CurrentTrustedRecoveryBootAssets)
+		if err != nil {
+			return nil, err
+		}
+		seq0 = append(seq0, bf.WithPath(p0))
+		seq1 = append(seq1, bf.WithPath(p1))
+	}
+
+	// add the recovery kernel
+	bf := bootChain[len(bootChain)-1]
+	seq0 = append(seq0, bf)
+	seq1 = append(seq1, bf)
+
+	if sequenceEqual(seq0, seq1) {
+		return [][]bootloader.BootFile{seq0}, nil
+	}
+
+	return [][]bootloader.BootFile{seq0, seq1}, nil
+}
+
+// kernelPathForRecoverySystem validates the kernel for the specified
+// recovery system and returns the path to the kernel snap.
+func kernelPathForRecoverySystem(recoverySystem string) (string, error) {
+	perf := timings.New(nil)
+	_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, recoverySystem, []snap.Type{snap.TypeKernel}, perf)
+	if err != nil {
+		return "", err
+	}
+	if len(snaps) != 1 {
+		return "", fmt.Errorf("cannot obtain recovery kernel snap")
+	}
+
+	return snaps[0].Path, nil
+}
+
+// runModeLoadSequences builds the run mode boot file list based on
+// the boot chains retrieved from the bootloaders and trusted assets
+// listed in modeenv.
+func runModeLoadSequences(rbl, bl bootloader.Bootloader, modeenv *Modeenv) ([][]bootloader.BootFile, error) {
+	trbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("recovery bootloader doesn't support trusted assets")
+	}
+
+	k0, k1, err := runModeKernelsFromModeenv(modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	bootChain, err := trbl.BootChain(bl, k0)
+	if err != nil {
+		return nil, err
+	}
+
+	seq0 := make([]bootloader.BootFile, 0, len(bootChain))
+	seq1 := make([]bootloader.BootFile, 0, len(bootChain))
```

Comment:
Why only 2 sequences? Shouldn't that be len(shim hashes) * len(recovery grub hashes) * len(run grub hashes) * len(current kernels)?

Reply:
I @cmatsuoka said he wants to do that as a 2nd step, OTOH I think we need to have good internal represenations in boot of the sequences because we also need code to decide whether to reseal at all